### PR TITLE
Second major commit

### PR DESCRIPTION
**Pull request by ahmad0596**

### CODE REVIEW

File: src/main/java/dawahdevs/qa/api/controller/QaApiController.java [hunk 1]
```diff
@@ -1,62 +1,64 @@
...
+		return qaApiManager.getUserById(userId);
+	}
+
+	@GetMapping("/authors")
+	public AllAuthorsResponse getAuthors() {
+		log.info("getting all authors");
+		return qaApiManager.getAuthors();
+	}
+
+	@GetMapping("/authors/{authorId}")
```
Comment (abubakir1997) [inline src/main/java/dawahdevs/qa/api/controller/QaApiController.java:59]: Is the convention to keep it plural instead of singular `/author/{authorId}`?

File: src/main/java/dawahdevs/qa/api/entity/DbAnswer.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+import org.hibernate.validator.constraints.Length;
+
+import lombok.Getter;
+import lombok.Setter;
+
+@Entity
+@Getter
+@Setter
+public class DbAnswer {
```
Comment (abubakir1997) [inline src/main/java/dawahdevs/qa/api/entity/DbAnswer.java]: Can we name it like the parent folder so `AnswerEntity`, `QuestionEntity`? Also this will be the exact name in the database which is fine but if you want you can add this annotation to specify the table name

 ```java
...
@Table(name = "answer")
public class AnswerEntity {...}
```

File: src/main/java/dawahdevs/qa/api/transformer/QaApiTransformer.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+		return allQuestionsResponse;
+	}
+
+	public QuestionByIdResponse transformQuestionById(final DbQuestion dbQuestion) {
+		QuestionByIdResponse questionByIdResponse = new QuestionByIdResponse();
+
+		optionalMap(dbQuestion, this::extractQuestion, questionByIdResponse::setQuestion);
+		// questionByIdResponse.setAnswers(answers);
+		// questionByIdResponse.setComments(comments);
```
Comment (abubakir1997) [inline src/main/java/dawahdevs/qa/api/transformer/QaApiTransformer.java]: Its part of best practice to never include commented out code, because it defeats the purpose of version control, if you want it you can look back to git history. The convention is to add a todo/fixme comment like this 
```
// TODO:  <TODO_MSG>
// FIXME:  <FIXME_MSG>
```

File: src/main/java/dawahdevs/qa/api/transformer/QaApiTransformer.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+import dawahdevs.qa.api.model.AuthorByIdResponse;
+import dawahdevs.qa.api.model.Question;
+import dawahdevs.qa.api.model.QuestionByIdResponse;
+import dawahdevs.qa.api.model.User;
+import dawahdevs.qa.api.model.UserByIdResponse;
+import lombok.AllArgsConstructor;
+
+@Component
+@AllArgsConstructor
+public class QaApiTransformer {
```
Comment (abubakir1997) [inline src/main/java/dawahdevs/qa/api/transformer/QaApiTransformer.java:30]: So I thought we mentioned that the only monolith would be the controller everything else will be broken into its own `Manager`, `Transformer`, `Validator`, etc. These could easily grow to none maintainable files.

File: src/test/java/dawahdevs/qa/api/controllerTest/QaApiControllerTest.java [hunk 5]
```diff
@@ -29,15 +35,65 @@
...
+		verify(qaApiManager, times(1)).getQuestions();
+	}
+
+	@Test
+	public void testGetQuestionById() throws Exception {
+		when(qaApiManager.getQuestionById(1)).thenReturn(new QuestionByIdResponse());
+
+		mockMvc.perform(get("/v1/questions/1")).andExpect(status().isOk()).andDo(print());
+
+		verify(qaApiManager, times(1)).getQuestionById(1);
```
Comment (abubakir1997) [inline src/test/java/dawahdevs/qa/api/controllerTest/QaApiControllerTest.java:60]: aren't you suppose to mock the `QuestionRepository` object only. I mean this doesn't add much testing value. This test is currently mocking the manager method then checking if it was mocked correctly. You should only mock what you cant control like a fetch request, an ORM, a repository query, third party libraries, etc. 

> same applies to the other tests

File: src/test/java/dawahdevs/qa/api/managerTest/QaApiManagerTest.java [hunk 6]
```diff
@@ -6,24 +6,33 @@
...
+//	private final QaApiTransformer qaApiTransformer;
+//
+//	@Mock
+//	private final QuestionRepository questionRepository;
+//
+//	@Mock
+//	private final UserRepository userRepository;
+//
+//	@Mock
+//	private final AuthorRepository authorRepository;
```
Comment (abubakir1997) [inline src/test/java/dawahdevs/qa/api/managerTest/QaApiManagerTest.java]: remove commented code please. it clutters the app code.

File: src/test/java/dawahdevs/qa/api/managerTest/QaApiManagerTest.java [hunk 7]
```diff
@@ -6,24 +6,33 @@
...
-	@Test
-	public void testGetQuestions() {
 	}
 
 	@Test
-	public void testGetQuestion() {
+	public void testGetQuestions() {
 	}
 
 }
```
Comment (abubakir1997) [inline src/test/java/dawahdevs/qa/api/managerTest/QaApiManagerTest.java:326]: In fact remove this entire test file since its empty. Dont add code unless it is needed now. This just adds more irrelevant code to maintain. Add future code in the future.

File: src/test/java/dawahdevs/qa/api/transformerTest/QaApiTransformerTest.java [hunk 8]
```diff
@@ -1,23 +1,127 @@
...
+		DbAuthor dbAuthor = new DbAuthor();
+
+		dbAuthor.setId(1);
+		dbAuthor.setName("Mohammad");
+		dbAuthor.setSchool(School.SHAFI);
+		dbAuthor.setSeniority(Seniority.EXPERT);
+		dbAuthor.setDefaultLang(Lang.EN);
+
+		return dbAuthor;
 	}
```
Comment (abubakir1997) [inline src/test/java/dawahdevs/qa/api/transformerTest/QaApiTransformerTest.java:158]: these `createDb...` methods should be in test utility files as they could be used by other test file and dont relate to the unit test.

**Review by abubakir1997 — CHANGES_REQUESTED** [review]
Left comments under **Files changed** tab

File: src/main/java/dawahdevs/qa/api/controller/QaApiController.java (hunk 1 above)
Comment (ahmad0596) [reply src/main/java/dawahdevs/qa/api/controller/QaApiController.java:59]: Yes, it is it is the collections/singleton pattern
https://restfulapi.net/resource-naming/

File: src/main/java/dawahdevs/qa/api/entity/DbAnswer.java (hunk 2 above)
Comment (ahmad0596) [reply src/main/java/dawahdevs/qa/api/entity/DbAnswer.java]: Done

File: src/main/java/dawahdevs/qa/api/transformer/QaApiTransformer.java (hunk 3 above)
Comment (ahmad0596) [reply src/main/java/dawahdevs/qa/api/transformer/QaApiTransformer.java]: Done

File: src/main/java/dawahdevs/qa/api/transformer/QaApiTransformer.java (hunk 4 above)
Comment (ahmad0596) [reply src/main/java/dawahdevs/qa/api/transformer/QaApiTransformer.java:30]: I will reduce code with generic Util methods in the future, because of the way data is returned from the backend, with question/answer/user/author data all in one body... seems sloppy transform each part of the payload in a different class...

As for the manager, code is so simplified that the Manager class will not be overloaded or hard to follow.

Sadly with the way our backend is giving us data we are forced to "look" monolithic sometimes but I def see you point!

**Review by abubakir1997 — APPROVED** [review]
Other than the questions I had on test mocks everything else is perfect, very good job!

File: src/test/java/dawahdevs/qa/api/controllerTest/QaApiControllerTest.java (hunk 5 above)
Comment (ahmad0596) [reply src/test/java/dawahdevs/qa/api/controllerTest/QaApiControllerTest.java:60]: So I definitely tested many different ways in older projects... one way being (like you mentioned) going all the way to the "DAO" and making sure I got there. 

HOWEVER the way I tested in this project as this is according to https://tanzu.vmware.com/application-modernization-recipes/testing/spring-boot-testing-best-practices .... You should ONLY test what this class controls so this test mocks a call and makes sure it goes to the proper manager method.... and if you think about it thats testing the exact and only functionality the controller controls. You will see that I only test what the Class itself is control over. So transformer simply asserts that I transformed the data payload properly, The validator simply asserts that my validation logic works and etc etc....

That being said, remember this is unit and coverage testing only... "Full (end to end) flow" testing can and will be handled elsewhere... for example, we can use postman for that. this way the test can also be excueted outside of maven build and etc... also against localhost or on the server
